Fixed #34574 -- Noted unexpected outcomes in autoescape/escape docs.

nessita opened the pull request.

nessita commented on docs/ref/templates/builtins.txt:
```diff
@@ -21,15 +21,6 @@ Controls the current auto-escaping behavior. This tag takes either ``on`` or
-The only exceptions are variables that are already marked as "safe" from
-escaping, either by the code that populated the variable, or because it has had
-the :tfilter:`safe` or :tfilter:`escape` filters applied.
-
```
I moved this paragraph to after the example because I think the reader may understand that there is a connection between these notes and the code that follows, but in reality the example is the simplest usage possible and (I think) it dilutes the seriousness of the two paragraphs.

nanis94 replied:
comment

felixxm replied:
@nanis94 Please stop, or I will block your account. You're spamming folks who watch this repository.

nessita commented on docs/ref/templates/builtins.txt:
```diff
@@ -38,6 +29,38 @@ Sample usage:
+..
+    Once https://code.djangoproject.com/ticket/34577 gets released, we should
+    add a sentence explaining how using :tfilter:`escapeseq` could be used to
+    properly implement the example above"
```
This is a comment to remember to extend this section.

felixxm replied:
Can we add `escapeseq` first? :thinking:

nessita replied:
Sure, updating now that the filter was added.

nanis94 commented:
nessita:ticket_34574 master

nessita commented:
@felixxm I rebased onto main and fixed conflicts. I also extended the docs referencing the `escapeseq` fitler. There are still 2 commits because, before rebasing/squashing, I wanted to confirm whether we want two isolated commits, one without any reference to `escapeseq` so we can backport it, and a second one with the `escapeseq` mention. Thanks!

felixxm commented on docs/ref/templates/builtins.txt:
```diff
@@ -1831,7 +1858,20 @@ For example, you can apply ``escape`` to fields when :ttag:`autoescape` is off:
+    outcomes if preceding filters mark a potentially unsafe string as safe due
+    to the lack of escaping caused by :ttag:`autoescape` being ``off``.
+
+    In such cases, similar to the note in the :tfilter:`safe` filter, chaining
```
The reference to `safe` sounds unnecessary to me, I'd chop it
```suggestion
    In such cases, chaining ...
```

felixxm commented on docs/ref/templates/builtins.txt:
```diff
@@ -1831,7 +1858,20 @@ For example, you can apply ``escape`` to fields when :ttag:`autoescape` is off:
+    to the lack of escaping caused by :ttag:`autoescape` being ``off``.
+
+    In such cases, similar to the note in the :tfilter:`safe` filter, chaining
+    ``escape`` would not re-escape strings that have already been marked as
```
Do we need a hyphen? (see ce586ed6931092d3a5f06df9031cdeb891793ddb) \cc @smithdc1
```suggestion
    ``escape`` would not reescape strings that have already been marked as
```

smithdc1 replied:
For `en-us` no hypen is required.

The AP Stylebook (a change in 2019) says to not to hypenate double "e" for "re-" and "pre-" combinations. Only exception would be if the sense was changed. e.g. resign (to quit) or re-sign (to sign again).

nessita replied:
Note that this change results in a spell check error:

```
WARNING: ref/templates/builtins.txt:1868: : Spell check: reescape:  would not reescape strings that have.
```

felixxm commented on docs/ref/templates/builtins.txt:
```diff
@@ -1831,7 +1858,20 @@ For example, you can apply ``escape`` to fields when :ttag:`autoescape` is off:
-To escape each element of a sequence, use the :tfilter:`escapeseq` filter.
+.. note::
+
+    As mentioned in the :ttag:`autoescape` section, when filters including
+    ``escape`` are chained together, it can result in unexpected
+    outcomes if preceding filters mark a potentially unsafe string as safe due
+    to the lack of escaping caused by :ttag:`autoescape` being ``off``.
```
Wrap at 79 chars:
```suggestion
    As mentioned in the :ttag:`autoescape` section, when filters including
    ``escape`` are chained together, it can result in unexpected outcomes if
    preceding filters mark a potentially unsafe string as safe due to the lack
    of escaping caused by :ttag:`autoescape` being ``off``.
```

felixxm commented on docs/ref/templates/builtins.txt:
```diff
@@ -1831,7 +1858,20 @@ For example, you can apply ``escape`` to fields when :ttag:`autoescape` is off:
+    ``escape`` would not re-escape strings that have already been marked as
+    safe.
+
+    This is specially important when using filters that would operate on
```
```suggestion
    This is specially important when using filters that operates on
```

felixxm reviewed (commented):
@nessita Thanks :+1: I left suggestions. We can keep two separate commits, but I'd change the commit message in the second one as it will not age well (`escapeseq` will not be new soonish), e.g. `Refs #34574 -- Mentioned escapeseq filter in escape/autoescape docs.`

smithdc1 commented on docs/ref/templates/builtins.txt:
```diff
@@ -38,6 +29,42 @@ Sample usage:
+:tfilter:`escape` filter to each variable.
+
+The only exceptions are variables that are already marked as "safe" from
+escaping: either by the code that populated the variable, or because it has had
+the :tfilter:`safe` or :tfilter:`escape` filters applied, or because it's the
+result of a previous filter that marked the string as "safe".
```
Either should be a choice from two options. There's three here. Maybe:

```suggestion
escaping. Variables could be marked as "safe" by the code which populated the variable, the use of :tfilter:`safe` or :tfilter:`escape` filters, or because it's the result of a previous filter that marked the string as "safe".
```

nessita replied:
I used David's suggestion with a minor tweak.

smithdc1 commented on docs/ref/templates/builtins.txt:
```diff
@@ -38,6 +29,42 @@ Sample usage:
+is because the filter chaining sequence executes first :tfilter:`join` on
+``my_list`` (without applying escaping to each item since ``autoescape`` is
+``off``), marking the result as safe. Subsequently, this safe result will be
+fed to :tfilter:`escape` filter, which, as documented, does not apply a second
```
I'm not sure about all these commas. Do we need the "as documented" part?

nessita replied:
Removed "as documented" and two commas.

smithdc1 commented on docs/ref/templates/builtins.txt:
```diff
@@ -38,6 +29,42 @@ Sample usage:
+filters have been applied). This is equivalent to manually applying the
+:tfilter:`escape` filter to each variable.
+
+The only exceptions are variables that are already marked as "safe" from
```
```suggestion
The only exceptions are variables already marked as "safe" from
```

smithdc1 commented on docs/ref/templates/builtins.txt:
```diff
@@ -38,6 +29,42 @@ Sample usage:
         {{ body }}
     {% endautoescape %}
 
+When auto-escaping is in effect, all content derived from variables have HTML
+escaping applied to it before placing the result into the output (but after any
+filters have been applied). This is equivalent to manually applying the
+:tfilter:`escape` filter to each variable.
```
Maybe:

```suggestion
When auto-escaping is in effect, all content derived from variables has HTML
escaping applied before placing the result into the output (but after any
filters are applied). This is equivalent to manually applying the
:tfilter:`escape` filter to each variable.
```

nessita replied:
Thanks! Applied (originally I did not touch this paragraph since I just moved it to after the example).

smithdc1 reviewed (commented):
Hi all. Looks good 👍 . As I was here I've left a few minor "nit-pick" comments.

nessita commented on docs/ref/templates/builtins.txt:
```diff
@@ -38,6 +29,42 @@ Sample usage:
+the :tfilter:`safe` or :tfilter:`escape` filters applied, or because it's the
+result of a previous filter that marked the string as "safe".
+
+It's important to note that within the scope of disabled auto-escaping,
```
Applied.

nessita commented on docs/ref/templates/builtins.txt:
```diff
@@ -38,6 +29,42 @@ Sample usage:
+result of a previous filter that marked the string as "safe".
+
+It's important to note that within the scope of disabled auto-escaping,
+chaining filters including :tfilter:`escape` could lead to unexpected (but
```
Changed.

nessita commented:
I believe I applied all suggestions and I kept the separation between two commits to ease backporting of the first one.

smithdc1 commented on docs/ref/templates/builtins.txt:
```diff
@@ -1831,7 +1858,19 @@ For example, you can apply ``escape`` to fields when :ttag:`autoescape` is off:
+    In such cases, chaining ``escape`` would not reescape strings that have
+    already been marked as safe.
+
+    This is specially important when using filters that operate on sequences,
```
```suggestion
    This is especially important when using filters that operate on sequences,
```

felixxm reviewed (approved):
@nessita Thanks :+1:
